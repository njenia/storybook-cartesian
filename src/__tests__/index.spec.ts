import cartesian, { choice, describedValue } from '../index'

const createStories = (s: any[]) => ({
  add: (title: string, story: any) => s.push({ title, ...story() })
})
const storyTitle = (props: any) => JSON.stringify(props)

const serializeStory = (props: any) => ({
  story: {
    props
  }
})
const createCartesian = (props: any) => {
  const s: any[] = []
  cartesian(createStories(s)).add(() => props, storyTitle, serializeStory)
  return s
}

describe('cartesian', () => {
  describe('choices', () => {
    it('empty prop', () => {
      expect(
        createCartesian({
          oneProp: choice()
        })
      ).toMatchSnapshot()
    })

    it('single prop', () => {
      expect(
        createCartesian({
          oneProp: choice(true)
        })
      ).toMatchSnapshot()
    })
    it('multi prop', () => {
      expect(
        createCartesian({
          oneProp: choice(true, false),
          twoProp: choice('', 'test')
        })
      ).toMatchSnapshot()
    })

    it('nested prop', () => {
      expect(
        createCartesian({
          oneProp: { story: 'yes', accept: choice(true, false) },
          twoProp: choice('', 'test')
        })
      ).toMatchSnapshot()
    })
    it('some invalid', () => {
      expect(
        createCartesian({
          oneProp: choice(true, false),
          twoProp: choice('', 'foobar', 'foobaz', 'test')
        })
      ).toMatchSnapshot()
    })
    it('described value', () => {
      expect(
        createCartesian({
          oneProp: choice(1, 2),
          twoProp: choice(describedValue('EMPTY TEXT', ''), 'test')
        })
      ).toMatchSnapshot()
    })
  })

  describe('vanilla', () => {
    it('empty prop', () => {
      expect(
        createCartesian({
          oneProp: []
        })
      ).toMatchSnapshot()
    })

    it('empty prop, complex', () => {
      expect(
        createCartesian({
          oneProp: [],
          twoProp: []
        })
      ).toMatchSnapshot()
    })

    it('single prop', () => {
      expect(
        createCartesian({
          oneProp: [true]
        })
      ).toMatchSnapshot()
    })

    it('empty', () => {
      expect(createCartesian({})).toMatchSnapshot()
    })

    it('multi prop', () => {
      expect(
        createCartesian({
          oneProp: ['one-1', 'one-2'],
          twoProp: ['two']
        })
      ).toMatchSnapshot()
    })
  })
})
