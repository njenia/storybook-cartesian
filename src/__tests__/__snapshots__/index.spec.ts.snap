// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cartesian choices described value 1`] = `
Array [
  Object {
    "story": Object {
      "props": Object {
        "oneProp": 1,
        "twoProp": "",
      },
    },
    "title": "{\\"oneProp\\":1,\\"twoProp\\":\\"\\",\\"__valueDescription\\":{\\"oneProp\\":null,\\"twoProp\\":\\"EMPTY TEXT\\"}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": 1,
        "twoProp": "test",
      },
    },
    "title": "{\\"oneProp\\":1,\\"twoProp\\":\\"test\\",\\"__valueDescription\\":{\\"oneProp\\":null,\\"twoProp\\":null}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": 2,
        "twoProp": "",
      },
    },
    "title": "{\\"oneProp\\":2,\\"twoProp\\":\\"\\",\\"__valueDescription\\":{\\"oneProp\\":null,\\"twoProp\\":\\"EMPTY TEXT\\"}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": 2,
        "twoProp": "test",
      },
    },
    "title": "{\\"oneProp\\":2,\\"twoProp\\":\\"test\\",\\"__valueDescription\\":{\\"oneProp\\":null,\\"twoProp\\":null}}",
  },
]
`;

exports[`cartesian choices empty prop 1`] = `Array []`;

exports[`cartesian choices multi prop 1`] = `
Array [
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
        "twoProp": "",
      },
    },
    "title": "{\\"oneProp\\":true,\\"twoProp\\":\\"\\",\\"__valueDescription\\":{\\"oneProp\\":null,\\"twoProp\\":null}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
        "twoProp": "test",
      },
    },
    "title": "{\\"oneProp\\":true,\\"twoProp\\":\\"test\\",\\"__valueDescription\\":{\\"oneProp\\":null,\\"twoProp\\":null}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": false,
        "twoProp": "",
      },
    },
    "title": "{\\"oneProp\\":false,\\"twoProp\\":\\"\\",\\"__valueDescription\\":{\\"oneProp\\":null,\\"twoProp\\":null}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": false,
        "twoProp": "test",
      },
    },
    "title": "{\\"oneProp\\":false,\\"twoProp\\":\\"test\\",\\"__valueDescription\\":{\\"oneProp\\":null,\\"twoProp\\":null}}",
  },
]
`;

exports[`cartesian choices nested prop 1`] = `
Array [
  Object {
    "story": Object {
      "props": Object {
        "oneProp": Object {
          "accept": true,
          "story": "yes",
        },
        "twoProp": "",
      },
    },
    "title": "{\\"twoProp\\":\\"\\",\\"oneProp\\":{\\"story\\":\\"yes\\",\\"accept\\":true},\\"__valueDescription\\":{\\"twoProp\\":null,\\"oneProp\\":null}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": Object {
          "accept": true,
          "story": "yes",
        },
        "twoProp": "test",
      },
    },
    "title": "{\\"twoProp\\":\\"test\\",\\"oneProp\\":{\\"story\\":\\"yes\\",\\"accept\\":true},\\"__valueDescription\\":{\\"twoProp\\":null,\\"oneProp\\":null}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": Object {
          "accept": false,
          "story": "yes",
        },
        "twoProp": "",
      },
    },
    "title": "{\\"twoProp\\":\\"\\",\\"oneProp\\":{\\"story\\":\\"yes\\",\\"accept\\":false},\\"__valueDescription\\":{\\"twoProp\\":null,\\"oneProp\\":null}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": Object {
          "accept": false,
          "story": "yes",
        },
        "twoProp": "test",
      },
    },
    "title": "{\\"twoProp\\":\\"test\\",\\"oneProp\\":{\\"story\\":\\"yes\\",\\"accept\\":false},\\"__valueDescription\\":{\\"twoProp\\":null,\\"oneProp\\":null}}",
  },
]
`;

exports[`cartesian choices single prop 1`] = `
Array [
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
      },
    },
    "title": "{\\"oneProp\\":true,\\"__valueDescription\\":{\\"oneProp\\":null}}",
  },
]
`;

exports[`cartesian choices some invalid 1`] = `
Array [
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
        "twoProp": "",
      },
    },
    "title": "{\\"oneProp\\":true,\\"twoProp\\":\\"\\",\\"__valueDescription\\":{\\"oneProp\\":null,\\"twoProp\\":null}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
        "twoProp": "foobar",
      },
    },
    "title": "{\\"oneProp\\":true,\\"twoProp\\":\\"foobar\\",\\"__valueDescription\\":{\\"oneProp\\":null,\\"twoProp\\":null}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
        "twoProp": "foobaz",
      },
    },
    "title": "{\\"oneProp\\":true,\\"twoProp\\":\\"foobaz\\",\\"__valueDescription\\":{\\"oneProp\\":null,\\"twoProp\\":null}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
        "twoProp": "test",
      },
    },
    "title": "{\\"oneProp\\":true,\\"twoProp\\":\\"test\\",\\"__valueDescription\\":{\\"oneProp\\":null,\\"twoProp\\":null}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": false,
        "twoProp": "",
      },
    },
    "title": "{\\"oneProp\\":false,\\"twoProp\\":\\"\\",\\"__valueDescription\\":{\\"oneProp\\":null,\\"twoProp\\":null}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": false,
        "twoProp": "foobar",
      },
    },
    "title": "{\\"oneProp\\":false,\\"twoProp\\":\\"foobar\\",\\"__valueDescription\\":{\\"oneProp\\":null,\\"twoProp\\":null}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": false,
        "twoProp": "foobaz",
      },
    },
    "title": "{\\"oneProp\\":false,\\"twoProp\\":\\"foobaz\\",\\"__valueDescription\\":{\\"oneProp\\":null,\\"twoProp\\":null}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": false,
        "twoProp": "test",
      },
    },
    "title": "{\\"oneProp\\":false,\\"twoProp\\":\\"test\\",\\"__valueDescription\\":{\\"oneProp\\":null,\\"twoProp\\":null}}",
  },
]
`;

exports[`cartesian vanilla empty 1`] = `Array []`;

exports[`cartesian vanilla empty prop 1`] = `Array []`;

exports[`cartesian vanilla empty prop, complex 1`] = `Array []`;

exports[`cartesian vanilla multi prop 1`] = `
Array [
  Object {
    "story": Object {
      "props": Object {
        "oneProp": "one-1",
        "twoProp": "two",
      },
    },
    "title": "{\\"oneProp\\":\\"one-1\\",\\"twoProp\\":\\"two\\",\\"__valueDescription\\":{\\"oneProp\\":null,\\"twoProp\\":null}}",
  },
  Object {
    "story": Object {
      "props": Object {
        "oneProp": "one-2",
        "twoProp": "two",
      },
    },
    "title": "{\\"oneProp\\":\\"one-2\\",\\"twoProp\\":\\"two\\",\\"__valueDescription\\":{\\"oneProp\\":null,\\"twoProp\\":null}}",
  },
]
`;

exports[`cartesian vanilla single prop 1`] = `
Array [
  Object {
    "story": Object {
      "props": Object {
        "oneProp": true,
      },
    },
    "title": "{\\"oneProp\\":true,\\"__valueDescription\\":{\\"oneProp\\":null}}",
  },
]
`;
